File: HelpfulSense/app/src/main/java/com/andruid/magic/helpfulsense/fragment/SettingsFragment.java
package com.andruid.magic.helpfulsense.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.EditTextPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;

import com.andruid.magic.helpfulsense.R;
import com.andruid.magic.helpfulsense.activity.IntroActivity;

import java.util.Objects;

public class SettingsFragment extends PreferenceFragmentCompat implements
        EditTextPreference.OnBindEditTextListener, Preference.OnPreferenceChangeListener {

    public static SettingsFragment newInstance(){
        return new SettingsFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.app_preferences);
        EditTextPreference threshPref = Objects.requireNonNull(findPreference(getString(R.string.pref_threshold)));
        threshPref.setOnBindEditTextListener(this);
        threshPref.setOnPreferenceChangeListener(this);
        threshPref.callChangeListener(PreferenceManager.getDefaultSharedPreferences(
                Objects.requireNonNull(getContext())).getString(getString(R.string.pref_threshold),
                getString(R.string.def_threshold)));

        EditTextPreference timesPref = Objects.requireNonNull(findPreference(getString(R.string.pref_time_stop)));
        timesPref.setOnBindEditTextListener(this);
        timesPref.setOnPreferenceChangeListener(this);
        timesPref.callChangeListener(PreferenceManager.getDefaultSharedPreferences(
                Objects.requireNonNull(getContext())).getString(getString(R.string.pref_time_stop),
                getString(R.string.def_time_stop)));
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_help, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId() == R.id.menu_help)
            startActivity(new Intent(getContext(), IntroActivity.class));
        return true;
    }

    @Override
    public void onBindEditText(@NonNull EditText editText) {
        editText.setInputType(InputType.TYPE_CLASS_NUMBER);
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        String key = preference.getKey();
        if(key.equals(getString(R.string.pref_threshold)) || key.equals(getString(R.string.pref_time_stop)))
            preference.setSummary((String)newValue);
        return true;
    }
}

File: HelpfulSense/app/src/main/java/com/andruid/magic/helpfulsense/ui/fragment/SettingsFragment.kt
package com.andruid.magic.helpfulsense.ui.fragment

import android.Manifest
import android.annotation.SuppressLint
import android.os.Bundle
import android.text.InputType
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.EditText
import androidx.navigation.fragment.findNavController
import androidx.navigation.ui.onNavDestinationSelected
import androidx.preference.EditTextPreference
import androidx.preference.EditTextPreference.OnBindEditTextListener
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.Preference.OnPreferenceChangeListener
import androidx.preference.PreferenceFragmentCompat
import com.andruid.magic.helpfulsense.R
import com.andruid.magic.helpfulsense.ui.util.buildInfoDialog
import com.andruid.magic.helpfulsense.ui.util.buildSettingsDialog
import com.andruid.magic.helpfulsense.util.getSimCards
import com.andruid.magic.locationsms.util.getSelectedSimSlot
import com.andruid.magic.locationsms.util.getShakeStopTime
import com.andruid.magic.locationsms.util.getShakeThreshold
import com.andruid.magic.locationsms.util.toast
import permissions.dispatcher.*

/**
 * Fragment to change settings
 */
@RuntimePermissions
class SettingsFragment : PreferenceFragmentCompat(), OnBindEditTextListener, OnPreferenceChangeListener {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.app_preferences)

        findPreference<EditTextPreference>(getString(R.string.pref_threshold))?.let {
            it.setOnBindEditTextListener(this)
            it.onPreferenceChangeListener = this
            it.callChangeListener(getShakeThreshold().toString())
        }
        findPreference<EditTextPreference>(getString(R.string.pref_time_stop))?.let {
            it.setOnBindEditTextListener(this)
            it.onPreferenceChangeListener = this
            it.callChangeListener(getShakeStopTime().toString())
        }
        initSimPreferenceWithPermissionCheck()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_help, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val navController = findNavController()
        return item.onNavDestinationSelected(navController) || super.onOptionsItemSelected(item)
    }

    override fun onBindEditText(editText: EditText) {
        editText.inputType = InputType.TYPE_CLASS_NUMBER
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any?): Boolean {
        when (preference.key) {
            getString(R.string.pref_threshold), getString(R.string.pref_time_stop) ->
                preference.summary = (newValue as String)
            getString(R.string.pref_sms_sim) -> {
                val listPreference = preference as ListPreference
                preference.summary = listPreference.entries[(newValue as String).toInt()]
            }
        }
        return true
    }

    /**
     * Initialize the SIM selection [ListPreference]
     */
    @SuppressLint("MissingPermission")
    @NeedsPermission(Manifest.permission.READ_PHONE_STATE)
    fun initSimPreference() {
        findPreference<ListPreference>(getString(R.string.pref_sms_sim))?.let {
            val simCards = getSimCards()
            val entries = simCards.map { sim -> sim.displayName }
            val values = simCards.map { sim -> sim.simSlotIndex.toString() }
            it.entries = entries.toTypedArray()
            it.entryValues = values.toTypedArray()
            it.onPreferenceChangeListener = this
            it.callChangeListener(getSelectedSimSlot().toString())
        }
    }

    @OnShowRationale(Manifest.permission.READ_PHONE_STATE)
    fun showRationale(request: PermissionRequest) {
        buildInfoDialog(R.string.phone_state_permission, request).show()
    }

    @OnPermissionDenied(Manifest.permission.READ_PHONE_STATE)
    fun showDenied() {
        toast("Denied permission")
    }

    @OnNeverAskAgain(Manifest.permission.READ_PHONE_STATE)
    fun showSettingsDialog() {
        buildSettingsDialog(R.string.phone_state_permission).show()
    }
}